File: daily-coding-problem/dcp-jvm/dcp-scala/src/main/scala/dcp/problems/_1_20/Problem15_RandElement.scala
package dcp.problems._1_20

import scala.util.Random

/**
 * This problem was asked by Facebook.
 *
 * Given a stream of elements too large to store in memory, pick a random element from the stream with
 * uniform probability.
 */

object Problem15_RandElement extends App {

  def solution(stream: LazyList[Int]): Int = {
    var randVal = 0
    val rand = new Random()
    for (i <- stream.indices) {
      if (i == 0) randVal = i
      else if (rand.between(1, i + 1) == 1) randVal = stream(i)
    }
    randVal
  }

  val stream: LazyList[Int] = LazyList.from(1 to 1000000)
  println(solution(stream))
  println(solution(stream))
  println(solution(stream))
}

File: daily-coding-problem/dcp-py/1_20/problem15_rand_element.py
"""
This problem was asked by Facebook.

Given a stream of elements too large to store in memory, pick a random element from the stream with uniform probability.

Reference: https://www.dailycodingproblem.com/blog/how-to-pick-a-random-element-from-an-infinite-stream/
"""
import random


def random_element(stream):
    random_val = None
    for i, elem in enumerate(stream):
        if i == 0:
            random_val = elem
        elif random.randint(1, i + 1) == 1:
            random_val = elem
    return random_val


big_stream = [i for i in range(1, 100)]

print(random_element(big_stream))
print(random_element(big_stream))
print(random_element(big_stream))
print(random_element(big_stream))
print(random_element(big_stream))
